tiktoken-rs/src/vendor_tiktoken.rs

```diff
@@ -541,6 +553,20 @@ impl CoreBPE {
             Err(e) => Err(anyhow!("Unable to decode into a valid UTF-8 string: {}", e)),
         }
     }
+
+    // tokenize a string and return the decoded tokens using the correct BPE model
+    // for example: "Hello world" -> ["Hello", " world"]
+    pub fn split_by_token_with_special_tokens(&self, text: &str) -> Result<Vec<String>> {
+        // first, encode the text using the BPE model
+        let encoded = self.encode_with_special_tokens(text);
+
+        let tokenized = self._decode_native_and_split(&encoded);
+
+        tokenized
+            .iter()
+            .map(|token| String::from_utf8(token.clone()).map_err(|e| anyhow!(e.to_string())))
+            .collect()
+    }
```

Comment:
Minor nit: With the subfunctions you defined, I see some unnecessary clones and Vec collections and memory allocations. Consider inlining `_decode_native_and_split` or changing the function signatures to avoid clones
```suggestion
    pub fn split_by_token_with_special_tokens(&self, text: &str) -> Result<Vec<String>> {
        // first, encode the text using the BPE model
        let encoded = self.encode_with_special_tokens(text);
        encoded
            .iter()
            .map(|token| {
                let token = self
                    .decoder
                    .get(token)
                    .unwrap_or_else(|| &self.special_tokens_decoder[token]);
                String::from_utf8(token.clone()).map_err(|e| anyhow!(e.to_string()))
            })
            .collect()
    }
```

Reply:
Disclaimer: I haven't thoroughly benchmarked the code so I'm not sure how big of an impact this would have on perf

Reply:
@zurawiki - What would you think about the function returning an iterator so that the downstream consumer only has to iterate once and has more flexibility? Something like:

```
pub struct DecodeNativeAndSplitIter<'a> {
    tokens: std::slice::Iter<'a, usize>,
    decoder: &'a HashMap<usize, Vec<u8>>,
    special_tokens_decoder: &'a HashMap<usize, Vec<u8>>,
}

impl<'a> Iterator for DecodeNativeAndSplitIter<'a> {
    type Item = &'a Vec<u8>;

    fn next(&mut self) -> Option<Self::Item> {
        self.tokens.next().map(|token| {
            self.decoder
                .get(token)
                .unwrap_or_else(|| &self.special_tokens_decoder[token])
        })
    }
}
```

My concern here is that it's a divergence from the patterns I see elsewhere in the codebase so we'd probably want to refactor other internal methods.

Reply:
I may have gone too far in the other direction, but now the function returns an iterator and no work is done until the user runs it. Also eliminated unnecessary clones. It makes the api a little less straightforward but I don't think it's too bad.